File: launcher/src/main/java/org/beangle/maven/launcher/downloader/RangeDownloader.java
package org.beangle.maven.launcher.downloader;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class RangeDownloader extends AbstractDownloader {

  private int threads = 20;
  private int step = 20 * 1024;

  private ExecutorService executor = Executors.newFixedThreadPool(threads);

  public RangeDownloader(String id, String url, String location) {
    super(id, url, location);
  }

  @Override
  protected void downloading() throws IOException {
    final URL resourceURL = new URL(url);
    URLConnection conn = resourceURL.openConnection();
    long startAt = System.currentTimeMillis();
    this.status = new DownloadStatus(conn.getContentLengthLong());
    if (this.status.total > Integer.MAX_VALUE) throw new RuntimeException(
        "Too large file.Using DefaultURLDownloader");
    int total = (int) this.status.total;

    final byte[] totalbuffer = new byte[total];
    int begin = 0;
    List<Callable<Integer>> tasks = new ArrayList<Callable<Integer>>();
    while (begin < this.status.total) {
      final int start = begin;
      final int end = ((start + step - 1) > total) ? total : (start + step - 1);
      tasks.add(new Callable<Integer>() {
        public Integer call() throws Exception {
          HttpURLConnection connection = (HttpURLConnection) resourceURL.openConnection();
          connection.setRequestProperty("RANGE", "bytes=" + start + "-" + end);
          InputStream input = connection.getInputStream();
          byte[] buffer = new byte[1024];
          int n = input.read(buffer);
          int next = start;

          while (-1 != n) {
            System.arraycopy(buffer, 0, totalbuffer, next, n);
            status.count.addAndGet(n);
            next += n;
            n = input.read(buffer);
          }
          close(input);
          return end;
        }
      });
      begin += step;
    }
    try {
      executor.invokeAll(tasks);
      executor.shutdown();
    } catch (InterruptedException e) {
      e.printStackTrace();
    }

    if (status.count.get() == status.total) {
      OutputStream output = new FileOutputStream(new File(location));
      output.write(totalbuffer, 0, total);
      close(output);
    } else {
      throw new RuntimeException("Download error");
    }
    finish(System.currentTimeMillis() - startAt);
  }

}

File: artifact/src/main/scala/org/beangle/maven/artifact/downloader/RangeDownloader.scala
/*
 * Beangle, Agile Development Scaffold and Toolkit
 *
 * Copyright (c) 2005-2015, Beangle Software.
 *
 * Beangle is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beangle is distributed in the hope that it will be useful.
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Beangle.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.beangle.maven.artifact.downloader

import java.io.{ File, FileOutputStream }
import java.net.{ HttpURLConnection, URL }
import java.util.concurrent.{ Callable, ExecutorService, Executors }
import org.beangle.commons.io.IOs

class RangeDownloader(name: String, url: String, location: String) extends AbstractDownloader(name, url, location) {

  var threads: Int = 20

  var step: Int = 20 * 1024

  var executor: ExecutorService = Executors.newFixedThreadPool(threads)

  protected override def downloading() {
    val resourceURL = new URL(url)
    val conn = resourceURL.openConnection()
    val startAt = System.currentTimeMillis()
    this.status = new Downloader.Status(conn.getContentLengthLong)
    if (this.status.total <= 0 || this.status.total > java.lang.Integer.MAX_VALUE) {
      throw new RuntimeException("Cannot download ${url} with size ${this.status.total}")
    }
    val total = this.status.total.toInt
    val totalbuffer = Array.ofDim[Byte](total)
    var begin = 0
    val tasks = new java.util.ArrayList[Callable[Integer]]
    while (begin < this.status.total) {
      val start = begin
      val end = if (((start + step - 1) > total)) total else (start + step - 1)
      tasks.add(new Callable[Integer]() {
        def call(): java.lang.Integer = {
          val connection = resourceURL.openConnection().asInstanceOf[HttpURLConnection]
          connection.setRequestProperty("RANGE", "bytes=" + start + "-" + end)
          val input = connection.getInputStream
          val buffer = Array.ofDim[Byte](1024)
          var n = input.read(buffer)
          var next = start
          while (-1 != n) {
            System.arraycopy(buffer, 0, totalbuffer, next, n)
            status.count.addAndGet(n)
            next += n
            n = input.read(buffer)
          }
          IOs.close(input)
          return end
        }
      })
      begin += step
    }
    try {
      executor.invokeAll(tasks)
      executor.shutdown()
    } catch {
      case e: InterruptedException => e.printStackTrace()
    }
    if (status.count.get == status.total) {
      val targetFile = new File(location)
      val output = new FileOutputStream(targetFile)
      output.write(totalbuffer, 0, total)
      targetFile.setLastModified(conn.getLastModified)
      IOs.close(output)
    } else {
      throw new RuntimeException("Download error")
    }
    finish(System.currentTimeMillis() - startAt)
  }

}
